c7n/filters/cost.py

```diff
@@ -0,0 +1,117 @@
+        'additionalProperties': False,
+        'required': ['type'],
+        'properties': {
+            'api_endpoint': {'type': 'string'},
+            'api_key': {'type': 'string'},
+            # 'currency': {'type': 'number'},
+            'quantity': {'type': 'number'},
+            'op': {'$ref': '#/definitions/filters_common/comparison_operators'},
+            'type': {'enum': ['cost']},
+            'value': {'type': 'number'},
+        },
+    }
+    schema_alias = True
+    ANNOTATION_KEY = "c7n:Cost"
+
+    def __init__(self, data, manager=None):
+        super().__init__(data, manager)
+        self.cache = manager._cache or NullCache({})
+        self.api_endpoint = data.get(
+            "api_endpoint",
+            os.environ.get("INFRACOST_API_ENDPOINT", "https://pricing.api.infracost.io"),
+        )
+        self.api_key = data.get("api_key", os.environ.get("INFRACOST_API_KEY"))
+
+    def get_permissions(self):
+        return ("ec2:DescribeInstances",)
+
+    def validate(self):
+        name = self.__class__.__name__
+        if self.api_endpoint is None:
+            raise ValueError("%s Filter requires Infracost pricing_api_endpoint" % name)
+
+        if self.api_key is None:
+            raise ValueError("%s Filter requires Infracost api_key" % name)
+        return super(Cost, self).validate()
+
+    def process_resource(self, resource, client, query):
+        price = self.get_price(resource, client, query)
+        op = self.data.get('operator', 'ge')
+        value = self.data.get('value', -1)
+        return OPERATORS[op](price["USD"], value)
+
+    def get_price(self, resource, client, query):
+        params = self.get_params(resource)
+        cache_key = str(params)
+
+        with self.cache:
+            price = self.cache.get(cache_key)
+            if not price:
+                price = self.invoke_infracost(client, query, params)
+                # TODO support configurable currency
+                price["USD"] = float(price["USD"]) * self.data.get("quantity", 1)
+                self.cache.save(cache_key, price)
+
+        resource[self.ANNOTATION_KEY] = price
+        return price
+
+    def invoke_infracost(self, client, query, params):
+        result = client.execute(query, variable_values=params)
+        self.log.info(f"Infracost {params}: {result}")
+        total = len(result["products"][0]["prices"])
+        if total > 1:
+            self.log.warning(f"Found {total} price options, expecting 1")
+        return result["products"][0]["prices"][0]
+
+    def process(self, resources, event=None):
+        transport = RequestsHTTPTransport(
+            url=self.api_endpoint + "/graphql",
+            headers={'X-Api-Key': self.api_key},
+            verify=True,
+            retries=5,
+        )
+        client = Client(transport=transport, fetch_schema_from_transport=True)
+        query = gql(self.get_query())
+        return [r for r in resources if self.process_resource(r, client, query)]
```

Comment:
this is making a lot of redundant api round trips.

ie say I have a 1000 m5.large in the same region, this is still going to do a thousand  round trips for now discernible purpose. the subclasses, should supporting grouping by the parameters to get the unique set across the resource pricing dimensions to minimize round trips.

Reply:
right, it would be better. Processing each resource here is to annotate the cost to resource dictionary. In fact, I was not intended to develop a filter because the initial requirement is to know a rough cost for our reports. For example, if a policy suggest that we should delete a unused EC2, we want to know how much it could save.

Reply:
I am going to keep the implementation unchanged as I need the annotation `side_effect`.